MoreLinq/SelectAsync.cs

```diff
@@ -0,0 +1,439 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2016 Atif Aziz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+#if !NO_ASYNC
+
+namespace MoreLinq
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Runtime.ExceptionServices;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Represents options for an asynchronous projection operation.
+    /// </summary>
+
+    public sealed class SelectAsyncOptions
```

Comment:
If equality is implemented, does the choice matter then? For non-changing options (each `With*` method returns `this` if nothing is changing), the reference optimises away copying that comes with value types.

Reply:
> This class appears to match value semantics

Do you say that on the sole basis that it is immutable? Building value semantics such as equality, while possible, may be a bit meaningless when `Scheduler` is a reference. In fact, I think `Scheduler` should even be a scheduler factory so that something like [`TaskScheduler.FromCurrentSynchronizationContext()`](https://docs.microsoft.com/en-gb/dotnet/api/system.threading.tasks.taskscheduler.fromcurrentsynchronizationcontext) is evaluated at this time of evaluation rather than at the time the query/enumerable/computation is setup.